### PR TITLE
Add support for Japanese BERT models by cl-tohoku

### PR DESCRIPTION
**Pull request by singletongue**
This PR adds new BERT models for Japanese text.
Details of the models can be found in [this repository](https://github.com/cl-tohoku/bert-japanese).

Since the way of tokenization is Japanese-specific, a new file is added for the tokenizers.
And, the locations of the files (models, configs, etc.) should be updated when the files are uploaded to S3. (How could I do this?)

### CODE REVIEW

**Comment by julien-c** [comment]
This looks great @singletongue. For the file upload, I will send you a way to upload to our S3, early next week, if that works for you.

**Comment by singletongue** [comment]
That will be great. Thank you!

File: transformers/tokenization_auto.py
Comment (LysandreJik) [inline transformers/tokenization_auto.py]: Auto models/tokenizer are used with checkpoint names, e.g.: `AutoModel.from_pretrained("bert-base-cased")`

^ it identifies `bert` in the checkpoint name and therefore initializes a Bert model from it.

If I understand correctly all the checkpoints you have have `bert-base-japanese` for prefix, which means that the auto detection won't work. Maybe replace it with `bert-base-japanese`?

**Comment by LysandreJik** [comment]
I could try it on my side, it seems to work well! 

I don't believe I've seen any benchmark on Japanese tasks in your repository, how did you evaluate the models? Did you use perplexity on portions of the Japanese wikipedia?

File: transformers/tokenization_auto.py
Comment (singletongue) [reply transformers/tokenization_auto.py]: Thank you for the review, @LysandreJik. I've fixed the AutoTokenizer.

**Comment by singletongue** [comment]
At the moment, there isn't an official publication regarding this work yet.

As a simple experiment, I applied the models to a Japanese document classification task using [livedoor news corpus](https://www.rondhuit.com/download.html#ldcc).
It is a 9-class classification, and I've confirmed that some of our models outperformed `bert-base-multilingual-cased` model.

| Model | Macro Precision | Macro Recall | Macro F1 |
| --- | --- | --- | --- |
| `bert-base-multilingual-cased` | 94.70 | 94.86 | 94.74 |
| `bert-base-japanese` | 95.64 | 95.58 | 95.60 |
| `bert-base-japanese-whole-word-masking` | 95.18 | 95.39 | 95.27 |
| `bert-base-japanese-char` | 94.32 | 93.96 | 94.09 |
| `bert-base-japanese-char-whole-word-masking` | 94.98 | 94.88 | 94.92 |

(Note that we expect the `char` models to be effective for SQuAD-like tasks, where span-level representation is needed. More comprehensive experiments are left for future work.)

Also, I've manually confirmed that our models predict reasonable tokens given a masked Japanese text.

Shall I add an example script for the classification task?

**Comment by LysandreJik** [comment]
Alright, thanks for clarifying. I don't think adding an example script is necessary.

This looks good to me but I'll wait for @thomwolf to chime in and review the PR before merging as it adds a few core functionalities (like the tokenizer).

**Comment by julien-c** [comment]
Hi @singletongue, the first step in uploading your weights to our S3 would be for you to create an account here: https://huggingface.co/join

The models will be namespaced under the chosen username or org name.

Can you let me know if you encounter any issue?

**Comment by singletongue** [comment]
Thank you for your guidance, @julien-c.

I've just created an account. Then I was told to use a CLI tool to log in and upload the model. However, I couldn't find the specified CLI in the repository.

**Comment by julien-c** [comment]
Yeah I should have been clearer, it's just that step 1 for now. What was clear or not clear in that page?

I will let you know here as soon as the CLI is operational.

**Comment by thomwolf** [comment]
This looks great to me, ok for merging.

Congratulations and thanks a lot @singletongue, this is an amazing work, both on preparing the data and training the model and on the super clean integration in the library. It's a masterpiece!

**Comment by thomwolf** [comment]
I think we add a test on the new tokenizer class though.

Do you think you could copy the file `tests/tokenization_bert_test` in a file `tests/tokenization_bert_japanese_test.py` and adapt a few English tests to Japanese specific characters @singletongue?

**Comment by singletongue** [comment]
Thank you for your review, @thomwolf.
I added a test for the tokenizers.

The test on CircleCI failed since the tokenizers need [MeCab](https://taku910.github.io/mecab/) and `mecab-python3` to be installed.
I've checked that the test run successfully on my local environment where the dependencies are installed.

**Comment by julien-c** [comment]
@singletongue The file upload CLI has been developed in https://github.com/huggingface/transformers/pull/2044 (soon to be merged to master)

Could you please upload your weights by doing:
```bash
git checkout cli_upload
pip install -e .
transformers-cli login
transformers-cli upload
```

Thanks!

**Comment by julien-c** [comment]
(For the tests requirements, maybe @thomwolf or @LysandreJik can chime in – I think you can either add a `pip install mecab-python3 ` to the config.yml file, or maybe create a specific test target for JP)

**Comment by singletongue** [comment]
Thank you, @julien-c. I've uploaded all the required files.
Also, I've updated `.circleci/config.yml` to successfully run tests on the tokenizers.

Thanks!

File: transformers/tests/utils.py
Comment (julien-c) [inline transformers/tests/utils.py]: cc @aaugustin

Comment (aaugustin) [reply transformers/tests/utils.py]: @julien-c My first instinct would to add the tokenizer test dependencies in the current test runs in Circle CI rather than setting up new test runs and adding the ability to skip these tests. Perhaps I'm missing a good reason why you did it that way?

I see that you recovered https://github.com/aaugustin/transformers/commit/17582458a30d0791ad4f2e18d7eb020f1dbc117f and made good use of it 👍

**Comment by thomwolf** [comment]
Ok LGTM

File: transformers/tests/utils.py
Comment (julien-c) [reply transformers/tests/utils.py]: Added a new build target because we expect to bundle other CJK languages in there in the future, and they come with a bunch of dependencies (system-wide, not just pip wide – cf the Python 2 target here) that are a PITA to have to install before running common tests.

Comment (julien-c) [reply transformers/tests/utils.py]: Added a new build target because we expect to bundle other CJK languages in there in the future, and they come with a bunch of dependencies (system-wide, not just pip wide – cf the Python 2 target here) that are a PITA to have to install before running common tests.

**Review by LysandreJik — APPROVED** [review]
Great!

**Pull request merged**
